README.md
region_analysis
===============

Region_analysis is a package derived and extended from region_analysis.pl in diffReps package. It is a utility to annotate the genomic intervals like the peak list of ChIP-seq or other interval lists from the genomic research. Now it supports human (hg19), mouse (mm9, mm10), and rat (rn4). New genomes will be added. Any question or suggestion is welcome!

Dependency:

  bedtools: [https://code.google.com/p/bedtools/](https://code.google.com/p/bedtools/)

  pybedtools: [https://github.com/daler/pybedtools](https://github.com/daler/pybedtools)

    If easy_install or pip is available, then:

      easy_install pybedtools

      or:

      pip isntall pybedtools

Usage: region_analysis.py [options]

Options:

  -h, --help            show this help message and exit

  -i INPUT_FILE_NAME, --input=INPUT_FILE_NAME

                        Input region file must assume the first 3 columns

                        contain (chr, start, end)

  -d ANNO_DB, --database=ANNO_DB

                        Choose database: refseq(default) or ensembl

  -r, --rhead           Whether the input file contains column header

  -g GENOME, --genome=GENOME

                        Choose genome: mm10(default), mm9, hg19, rn4.

  -v, --version

                        Version of Region_Analysis package

Output:

  -.annotated: the one-to-one output list, only the annotation entry whose TSS is nearest to the inquiry interval kept.

  -.full.annotated: all hit entries are kept.

  -.full.annotated.json: the json format output of -.full.annotated.
  
  ProximalPromoter: 	+/- 250bp of TSS

  Promoter1k: 	+/- 1kbp of TSS

  Promoter3k: 	+/- 3kbp of TSS

  Genebody: 	Anywhere between a gene's promoter and up to 1kbp downstream of the TES.

  Genedeserts: 	Genomic regions that are depleted with genes and are at least 1Mbp long.

  Pericentromere: 	Between the boundary of a centromere and the closest gene minus 10kbp of that gene's regulatory region.

  Subtelomere: 	Similary defined as pericentromere.

  OtherIntergenic: 	Any region that does not belong to the above categories.

Testing with examples:

region_analysis.py -i example/test_without_header.bed -g mm10 -d ensembl

region_analysis.py -i example/test_with_header.bed -g mm10 -d ensembl -r


CHANGES.txt
CHANGES
=======

v0.1.1
There is a bug in midpoint function in pybedtools 0.6.3, so an alternative function was used.

v0.1
The first release. With the annotations of hg19, mm9, mm10, and rn4.

tests/test_region_analysis.py
import unittest
import subprocess
import os
import regionanalysis

class TestRA(unittest.TestCase):

    def setUp(self):
        self.module_dir = os.path.dirname(
            os.path.realpath(regionanalysis.__file__))
        self.example_dir = os.path.join(self.module_dir, "example/")
        self.exception_dir = os.path.join(self.module_dir, "exceptions/")
        print(self.module_dir)

    def test_with_header(self):
        input_example = os.path.join(self.example_dir, "test_with_header.bed")
        cmds = ["region_analysis.py", "-d", "ensembl",
                "-g", "mm10", "-i", input_example, "-r"]
        p = subprocess.Popen(cmds, stdout=subprocess.PIPE, bufsize=1)
        stdout, stderr = p.communicate()
        test_output = open(input_example + ".annotated").read()
        test_ref = open(input_example + ".annotated.out").read()
        self.assertMultiLineEqual(test_output, test_ref)
        test_output = open(input_example + ".full.annotated").read()
        test_ref = open(input_example + ".full.annotated.out").read()
        self.assertMultiLineEqual(test_output, test_ref)
        test_output = open(input_example + ".full.annotated.json").read()
        test_ref = open(input_example + ".full.annotated.json.out").read()
        self.assertMultiLineEqual(test_output, test_ref)

    def test_without_header(self):
        input_example = os.path.join(
            self.example_dir, "test_without_header.bed")
        cmds = ["region_analysis.py", "-d", "ensembl",
                "-g", "mm10", "-i", input_example]
        p = subprocess.Popen(cmds, stdout=subprocess.PIPE, bufsize=1)
        stdout, stderr = p.communicate()
        test_output = open(input_example + ".annotated").read()
        test_ref = open(input_example + ".annotated.out").read()
        self.assertMultiLineEqual(test_output, test_ref)
        test_output = open(input_example + ".full.annotated").read()
        test_ref = open(input_example + ".full.annotated.out").read()
        self.assertMultiLineEqual(test_output, test_ref)
        test_output = open(input_example + ".full.annotated.json").read()
        test_ref = open(input_example + ".full.annotated.json.out").read()
        self.assertMultiLineEqual(test_output, test_ref)

    def test_exceptions(self):
        # in subprocess.Popen, shell=True, executable="bash" were used, because in UBUNTU, /bin/sh is not bash!
        # null input
        null_msg = "Please assign proper input file!"
        cmds = ["region_analysis.py"]
        p = subprocess.Popen(
            cmds, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
            bufsize=1, universal_newlines=True, shell=False)
        stdout, stderr = p.communicate()
        self.assertEqual(null_msg, stderr)
        del p, stdout, stderr
        # try different input files
        genomes = ["dumb", "mm10", "mm10"]
        msgs = ["dumb not in the genome database!",
                "Error in input file! Please check the format!",
                "Error in input file! Please check the format!"]
        for i in range(0, 3):
            input_example = os.path.join(self.exception_dir, "%d.bed"%i)
            cmds = ["region_analysis.py", "-d", "ensembl",
                    "-g", genomes[i], "-i", input_example]
            print(cmds)
            p = subprocess.Popen(
                cmds, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                bufsize=1, universal_newlines=True, shell=False)
            stdout, stderr = p.communicate()
            print(stdout)
            print(stderr)
            self.assertEqual(msgs[i], stderr)
            del p, stdout, stderr


if __name__ == '__main__':
    unittest.main()


regionanalysis/__init__.py
__version__ = "0.1.1"

def midpoint(cur_input):
    """
    Calculate the midpoint of the genomic interval.
    cur_input: a line of input, the first 3 columns are the intervals in BED format.
    """
    mid_input = cur_input.split("\t")
    midpoint = (int(mid_input[1]) + int(mid_input[2]))/2
    mid_input[1] = str(midpoint)
    mid_input[2] = str(midpoint + 1)
    return "\t".join(mid_input)

def getDis2TSS(anno_db, cur_input, col_no_input):
    """
    Calculate the distance to TSS and decide the annotation feature of the entry.
    """
    # calculate TSS and TES based on strands.
    # if strand is not "-" then it will be treated as "+".
    if cur_input[col_no_input + 5] != "-":
        TSS = int(cur_input[col_no_input + 10])
        TES = int(cur_input[col_no_input + 11])
        cur_input[col_no_input + 1] = str(int(TSS))
        cur_input[col_no_input + 2] = str(int(TES))
        Dis2TSS = int(cur_input[1]) - TSS
        Dis2TES = int(cur_input[1]) - TES
    else:
        TSS = int(cur_input[col_no_input + 11])
        TES = int(cur_input[col_no_input + 10])
        cur_input[col_no_input + 1] = str(int(TES))
        cur_input[col_no_input + 2] = str(int(TSS))
        Dis2TSS = TSS - int(cur_input[1])
        Dis2TES = TES - int(cur_input[1])
    if abs(Dis2TSS) <= 250:
        Pos = "ProximalPromoter"
    elif abs(Dis2TSS) <= 1000:
        Pos = "Promoter1k"
    elif abs(Dis2TSS) <= 3000:
        Pos = "Promoter3k"
    else:
        Pos = "genebody"
    if anno_db == "ensembl":  # output is gid
        cur_output = [
            cur_input[col_no_input+3], cur_input[col_no_input+5], cur_input[col_no_input+6],
            cur_input[col_no_input+10], cur_input[col_no_input+11], Pos, str(Dis2TSS),
            cur_input[col_no_input+9]]
    else:  # output is gene symbol
        cur_output = [cur_input[col_no_input+4], cur_input[col_no_input+5], cur_input[col_no_input+6],
            cur_input[col_no_input+10], cur_input[col_no_input+11], Pos, str(Dis2TSS),
            cur_input[col_no_input+9]]
    return (cur_output, Dis2TSS, Dis2TES)

def getBestHit(anno_db, col_no_input, GB_entry, gd_entry, st_entry, pc_entry):
    """
    format all hits and get the best hit that is neartest to TSS.
    """
    best_hit, cur_output, Dis2TSS, Dis2TES = (None, None, None, None)
    formatted = []

    for i in GB_entry:
        # discard null hit of genebody entry.
        if not ((i[col_no_input] == ".") and (i[col_no_input + 1] == "-1")):
            cur_output, Dis2TSS, Dis2TES = getDis2TSS(
                anno_db, i, col_no_input)
            if (best_hit is None) or (abs(int(best_hit[6])) > abs(Dis2TSS)):
                best_hit = cur_output
            formatted.append(cur_output)
    if gd_entry != "0":
        cur_output = ["NA", "NA", ".", "NA", "NA", "Genedesert", "NA", "No_anno"]
        if (best_hit is None):
            best_hit = cur_output
        formatted.append(cur_output)
    if st_entry != "0":
        cur_output = ["NA", "NA", ".", "NA", "NA", "Subtelomere", "NA", "No_anno"]
        if (best_hit is None):
            best_hit = cur_output
        formatted.append(cur_output)
    if pc_entry != "0":
        cur_output = ["NA", "NA", ".", "NA", "NA",
                      "Pericentromere", "NA", "No_anno"]
        if (best_hit is None):
            best_hit = cur_output
        formatted.append(cur_output)
    if best_hit is None:
        cur_output = ["NA", "NA", ".", "NA", "NA",
                      "Otherintergenic", "NA", "No_anno"]
        best_hit = cur_output
        formatted.append(cur_output)
    return (formatted, best_hit)

region_analysis.py
#! /usr/bin/env python

import os
import sys
import json
from itertools import groupby
from optparse import OptionParser
import pybedtools
import pybedtools.featurefuncs


def main():
    opt_parser = OptionParser()
    opt_parser.add_option('-i', '--input', action='store',
                          dest='input_file_name',
                          help='Input region file must assume the first 3 columns contain (chr, start, end)')
    opt_parser.add_option('-d', '--database', action='store',
                          dest='anno_db', help='Choose database: refseq(default) or ensembl',
                          default='refseq')
    opt_parser.add_option('-r', '--rhead', action='store_true',
                          dest='rhead', help='Whether the input file contains column header', default=False)
    opt_parser.add_option('-g', '--genome', action='store',
                          dest='genome', help='Choose genome: mm10(default)',
                          default='mm10')
    try:    
        (options, args) = opt_parser.parse_args(sys.argv)
        script_dir = os.path.dirname(os.path.realpath(__file__))
        db_path = os.path.join(script_dir, "database/")
        input_file_name = options.input_file_name
        anno_db = options.anno_db
        rhead = options.rhead
        genome = options.genome
        if (input_file_name is None) or (len(input_file_name)==0):
            raise SystemExit
    except SystemExit:
        print("Please assign proper input file!")
        opt_parser.print_help()
        return 1

    # create a tmp bed file with index column
    in_f = file(input_file_name)
    input_filtered = [ line  for line in in_f  if not line.lstrip().startswith("#") ] # filter the comment lines
    # if there is header, store it and remove it from the query BED.
    if rhead == True:
        headlineL = input_filtered[0].strip().split("\t")
        del input_filtered[0]
    input_indexed = [ '%s\t%d\n' % (line.strip(), i) for i, line in enumerate(input_filtered) ] # add index column to the bed lines
    in_f.close()

    # read all annotations into a dictionary, for the further output.
    anno_bed = os.path.join(
        db_path, genome + "." + anno_db + ".biotype_region_ext.bed")

    # use saveas() to convert the BedTool objects to file-based objects,
    # so they could be used multiple times.
    # When debug, we may use saveas("tss.tmp"), and the output of bedtools
    # could be saved.
    pybedtools.set_tempdir("./")
    anno = pybedtools.BedTool(anno_bed).saveas()
    gd = pybedtools.BedTool(
        os.path.join(db_path, genome + "_geneDesert.bed")).saveas()
    pc = pybedtools.BedTool(
        os.path.join(db_path, genome + "_pericentromere.bed")).saveas()
    st = pybedtools.BedTool(
        os.path.join(db_path, genome + "_subtelomere.bed")).saveas()

    # load the input intervals to be annotated
    input_bed = pybedtools.BedTool(
        "".join(input_indexed), from_string=True).saveas()
    list_input = [x.fields[:] for x in input_bed]
    col_no_input = input_bed.field_count()
    # get the midpoint of the intervals
    input_bed_mid = input_bed.each(pybedtools.featurefuncs.midpoint).saveas()

    # intersectBed with annotations
    input_GB = input_bed_mid.intersect(anno, wao=True).saveas()
    list_GB = [x.fields[:] for x in input_GB]
    input_gd = input_bed_mid.intersect(gd, c=True, f=0.5).saveas()
    list_gd = [x.fields[col_no_input + 0] for x in input_gd]
    input_pc = input_bed_mid.intersect(pc, c=True, f=0.5).saveas()
    list_pc = [x.fields[col_no_input + 0] for x in input_pc]
    input_st = input_bed_mid.intersect(st, c=True, f=0.5).saveas()
    list_st = [x.fields[col_no_input + 0] for x in input_st]

    # groupby the intersectBed results based on the index column
    input_idx = key = lambda s: s[col_no_input - 1]
    GB_dict = {}
    for key, GB_hits in groupby(list_GB, key=input_idx):
        GB_dict[key] = list(v for v in GB_hits)

    def getBestHit(anno_db, col_no_input, GB_entry, gd_entry, st_entry, pc_entry):
        """
        format all hits and get the best hit that is neartest to TSS
        """
        best_hit, cur_output, Dis2TSS, Dis2TES = (None, None, None, None)
        formatted = []

        def getDis2TSS(anno_db, cur_input, col_no_input):
            """
            Calculate the distance to TSS and decide the annotation feature of the entry
            """
            # calculate TSS and TES based on strands
            # if strand is not "-" then it will be treated as "+"
            if cur_input[col_no_input + 5] != "-":
                TSS = int(cur_input[col_no_input + 10])
                TES = int(cur_input[col_no_input + 11])
                cur_input[col_no_input + 1] = str(int(TSS))
                cur_input[col_no_input + 2] = str(int(TES))
                Dis2TSS = int(cur_input[1]) - TSS
                Dis2TES = int(cur_input[1]) - TES
            else:
                TSS = int(cur_input[col_no_input + 11])
                TES = int(cur_input[col_no_input + 10])
                cur_input[col_no_input + 1] = str(int(TES))
                cur_input[col_no_input + 2] = str(int(TSS))
                Dis2TSS = TSS - int(cur_input[1])
                Dis2TES = TES - int(cur_input[1])
            if abs(Dis2TSS) <= 250:
                Pos = "ProximalPromoter"
            elif abs(Dis2TSS) <= 1000:
                Pos = "Promoter1k"
            elif abs(Dis2TSS) <= 3000:
                Pos = "Promoter3k"
            else:
                Pos = "genebody"
            if anno_db == "ensembl":  # output is gid
                cur_output = [
                    cur_input[col_no_input+3], cur_input[col_no_input+6], cur_input[col_no_input+5],
                    cur_input[col_no_input+10], cur_input[col_no_input+11], Pos, str(Dis2TSS),
                    cur_input[col_no_input+9]]
            else:  # output is gene symbol
                cur_output = [cur_input[col_no_input+4], cur_input[col_no_input+6], cur_input[col_no_input+5],
                    cur_input[col_no_input+10], cur_input[col_no_input+11], Pos, str(Dis2TSS),
                    cur_input[col_no_input+9]]
            return (cur_output, Dis2TSS, Dis2TES)

        for i in GB_entry:
            # discard null hit of genebody entry
            if not ((i[col_no_input] == ".") and (i[col_no_input + 1] == "-1")):
                cur_output, Dis2TSS, Dis2TES = getDis2TSS(
                    anno_db, i, col_no_input)
                if (best_hit is None) or (abs(int(best_hit[6])) > abs(Dis2TSS)):
                    best_hit = cur_output
                formatted.append(cur_output)
        if gd_entry != "0":
            cur_output = ["", "", "", "", "", "Genedesert", "NA", "No_anno"]
            if (best_hit is None):
                best_hit = cur_output
            formatted.append(cur_output)
        if st_entry != "0":
            cur_output = ["", "", "", "", "", "Subtelomere", "NA", "No_anno"]
            if (best_hit is None):
                best_hit = cur_output
            formatted.append(cur_output)
        if pc_entry != "0":
            cur_output = ["", "", "", "", "",
                          "Pericentromere", "NA", "No_anno"]
            if (best_hit is None):
                best_hit = cur_output
            formatted.append(cur_output)
        if best_hit is None:
            cur_output = ["", "", "", "", "",
                          "Otherintergenic", "NA", "No_anno"]
            best_hit = cur_output
            formatted.append(cur_output)
        return (formatted, best_hit)

    output_file_best = file(input_file_name + ".annotated", "w")
    output_file = file(input_file_name + ".full.annotated", "w")
    output_file_json = file(input_file_name + ".full.annotated.json", "w")
    # Output the header
    if rhead == True:
        output_file.write("\t".join(
            headlineL + ["GName", "TName", "Strand", "TSS", "TES", "Feature", "D2TSS", "Biotype"]) + "\n")
        output_file_best.write("\t".join(
            headlineL + ["GName", "TName", "Strand", "TSS", "TES", "Feature", "D2TSS", "Biotype"]) + "\n")
    # write to the output: input.bed.annotated, input.bed.full.annotated
    json_dict = {}
    for i in range(0, len(input_bed)):
        output_lineL = list_input[i][:-1]  # original input line
        json_dict[str(i)] = {}
        json_dict[str(i)]["query_interval"] = output_lineL
        formatted, best_hit = getBestHit(
            anno_db, col_no_input, GB_dict[str(i)], list_gd[i], list_st[i], list_pc[i])
        output_file_best.write("\t".join(output_lineL + best_hit) + "\n")
        json_dict[str(i)]["best_hit"] = best_hit
        for j in formatted:
            output_file.write("\t".join(output_lineL + j) + "\n")
        json_dict[str(i)]["all_hits"] = formatted
    output_file_best.close()
    output_file.close()
    json.dump(json_dict, output_file_json, sort_keys=True, indent=2)
    output_file_json.close()
    pybedtools.cleanup()
    return 0

#-------------------------------------------------------------------------
if __name__ == '__main__':
    main()
#-------------------------------------------------------------------------
# EOF
